source/opt/eliminate_dead_input_components_pass.cpp

```diff
@@ -55,23 +63,37 @@ Pass::Status EliminateDeadInputComponentsPass::Process() {
     if (ptr_type == nullptr) {
       continue;
     }
+    auto sclass = ptr_type->storage_class();
     if (output_instead_) {
-      if (ptr_type->storage_class() != spv::StorageClass::Output) {
+      if (sclass != spv::StorageClass::Output) {
         continue;
       }
     } else {
-      if (ptr_type->storage_class() != spv::StorageClass::Input) {
+      if (sclass != spv::StorageClass::Input) {
         continue;
       }
     }
-    const analysis::Array* arr_type = ptr_type->pointee_type()->AsArray();
+    // For tesc, or input variables in tese or geom shaders,
+    // only optimize arrays and skip the first level of arrayness.
```

Comment:
Can you explain why? IMO a comment should not describe what the code does, but why the code is doing something.

Reply:
In general, these particular variables MUST be arrays. The outer array is called the "per-vertex-array". It is completely orthogonal to this analysis/optimization and can and really must be ignored. This is a somewhat tricky and confusing aspect of input/output variables in these shaders. Please refer to the GLSL spec and search for per-vertex-array.

I will improve the comment per this explanation.

Reply:
For example, if the vertex shader looks like this:

```
#version 450

layout (location = 0) out Vertex
{
    vec4 Cd;
    vec2 uv;
} oVert;

layout (location = 0) in vec4 P;

void main()
{
    oVert.Cd = P;
}
```

the following tesc shader will match:

```
#version 450

layout (vertices = 4) out;

layout (location = 0) in Vertex
{
    vec4 Cd;
    vec2 uv;
} iVert[];

layout (location = 0) out vec4 position[4];

void main()
{
    vec4 pos = iVert[gl_InvocationID].Cd;
    position[gl_InvocationID] = pos;
}
```

Note oVert is scalar in the vert shader, but an array in the tesc shader. See also 15.1.3 in the Vulkan spec.